### DISCUSSION TITLE
Updating items with XOR constraints

### DISCUSSION THREAD

**Discussion by Christian-Barnowski** [prisma/prisma#8462] [Q&A] [answered]
This has some similarities with [this question](https://github.com/prisma/prisma/discussions/3495) but concerns only a single item.

We have an `Item` that must reference either a `Person` or an `Organization` but not both. The model looks somewhat like this:
```Prisma
model Item {
  id              String        @id

  personId        String?       @db.Uuid
  organizationId  String?       @db.Uuid

  organization    Organization?     @relation(fields: [organizationId], references: [id])
  person          Person?    @relation(fields: [personId], references: [id])
}
```

To make sure one, and only one, reference is set we currently use a check constraint like this:
```SQL
ALTER TABLE "Item"
ADD CONSTRAINT "personOrOrganization"
  CHECK (("personId" IS NOT NULL AND "organizationId" IS NULL)
  OR ("personId" IS NULL AND "organizationId" IS NOT NULL));
```

This works fine for creating a new `Item` or, let's say, change a referenced `Person` to another one.
However sometimes we have the need to remove one type of reference and add the other. In this case `prisma.update()` fails because of the check constraint as it's doing one after the other.

```Javascript
prisma.item.update({
      where: {
        id: id,
      },
      data: {
        person: {
          disconnect: true,
        },
        organization: {
          connect: {
            id: orgId,
          },
        },
      },
    });
```

```
prisma:query BEGIN
prisma:query SELECT "public"."Organization"."id" FROM "public"."Organization" WHERE "public"."Organization"."id" = $1 OFFSET $2
prisma:query SELECT "public"."Item"."id" FROM "public"."Item" WHERE "public"."Item"."id" = $1
prisma:query UPDATE "public"."Item" SET "organizationId" = $1 WHERE "public"."Item"."id" IN ($2)
prisma:query ROLLBACK
```

At the moment we get around this by executing the update as raw SQL. This works so far but we were wondering if there's something we're overlooking - a prisma feature that would allow us to get rid of the raw SQL or maybe a recommended approach different to ours that achieves the same thing?

**Comment by ryands17** [comment, marked as answer]
@Christian-Barnowski 👋 

You could do this to update it in a single call so that it doesn't violate your `check` constraint:

```ts
await prisma.item.update({
    where: { id: id },
    data: { personId: null, organizationId: orgId },
})
```

**Reply by Christian-Barnowski** [reply]
Nice, thank you very much!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
